tool/gravity/cli/clusterupdate.go

```diff
@@ -237,7 +237,9 @@ func executeUpdatePhaseForOperation(env *localenv.LocalEnvironment, environ Loca
 func executeOrForkPhase(env *localenv.LocalEnvironment, updater updater, params PhaseParams, operation ops.SiteOperation) error {
 	if params.isResume() {
 		if err := verifyOrDeployAgents(env); err != nil {
-			return trace.Wrap(err)
+			// Continue operation in case gravity-site or etcd is down. In these
+			// cases the agent status may not be retrievable.
+			log.WithError(err).Warn("Failed to verify or deploy agents.")
```

Comment:
Technically we can test whether agents are active using the static database that we create and keep as the source of truth during the upgrade on each node. This is a more involved change though as one would have to rethink the way we query the cluster state to get the list of servers. 

 * `collectAgentStatus` only requires the operator to get the list of servers which is available in the static database.
 * `rpcDeployAgent` does rely on `localenv.LocalEnvironment` and `localenv.ClusterEnvironment` for credentials handling but I believe this is a design mistake. If we want to keep the upgrade operation resilient then pulling (and even upserting as I've seen!) credentials from/to the cluster package service is not a viable option. The secrets should continue to be extracted as files and used only as such during the operation (they are not both - extracted as files and pulled from the cluster package service).

When the operation is initialized, it performs the following:
  1. create operation/plan
  1. sync operation plan into a local database file which is kept as a source of truth
  1. extract RPC creds on each node for the agent
  1. deploy agents
Then, depending on the operational mode, it either starts executing the plan or pauses (e.g. manual mode)
Given all these prerequisites, it should be possible to verify the state of agents and redeploy agents if necessary at any given time using just the static database.

I'd really like that we take some time to consider this carefully and make improvements to the current state of flux.

Reply:
I'm not sure how we can redeploy agents without gravity-site running because agents deployment is done through teleport.

Also, maybe we can redesign this somehow in master if we want, but when backporting to LTS branches let's keep the changeset to a minimum and not make any redesigns without a good reason to reduce risk.